nvflare/fuel/f3/drivers/aio_conn.py

```diff
@@ -0,0 +1,142 @@
+from nvflare.fuel.f3.drivers.driver_params import DriverParams
+from nvflare.fuel.f3.sfm.prefix import PREFIX_LEN, Prefix
+from nvflare.fuel.hci.security import get_certificate_common_name
+
+log = logging.getLogger(__name__)
+
+MAX_FRAME_SIZE = 1024 * 1024 * 1024
+
+
+class AioConnection(Connection):
+    def __init__(
+        self,
+        connector: ConnectorInfo,
+        aio_ctx: AioContext,
+        reader: StreamReader,
+        writer: StreamWriter,
+        secure: bool = False,
+    ):
+        super().__init__(connector)
+        self.reader = reader
+        self.writer = writer
+        self.aio_ctx = aio_ctx
+        self.closing = False
+        self.secure = secure
+        self.conn_props = self._get_aio_properties()
+
+    def get_conn_properties(self) -> dict:
+        return self.conn_props
+
+    def close(self):
+        self.closing = True
+
+        if not self.writer:
+            return
+
+        self.writer.close()
+        self.aio_ctx.run_coro(self.writer.wait_closed())
+
+    def send_frame(self, frame: BytesAlike):
+        try:
+            self.aio_ctx.run_coro(self._async_send_frame(frame))
+        except Exception as ex:
+            log.error(f"Error calling send coroutine for connection {self}: {ex}")
+
+    async def read_loop(self):
+        try:
+            while not self.closing:
+                frame = await self._async_read_frame()
+                self.process_frame(frame)
+
+        except IncompleteReadError:
+            if log.isEnabledFor(logging.DEBUG):
+                closer = "locally" if self.closing else "by peer"
+                log.debug(f"Connection {self} is closed {closer}")
+        except CancelledError as error:
+            log.debug(f"Connection {self} is closed by peer: {error}")
+        except Exception as ex:
+            log.error(f"Read error for connection {self}: {ex}")
+
+    # Internal methods
+
+    async def _async_send_frame(self, frame: BytesAlike):
+        try:
+            self.writer.write(frame)
+            await self.writer.drain()
+        except Exception as ex:
+            log.error(f"Error sending frame for connection {self}: {ex}")
+
+    async def _async_read_frame(self):
+
+        prefix_buf = await self.reader.readexactly(PREFIX_LEN)
+        prefix = Prefix.from_bytes(prefix_buf)
+
+        # Prefix only message
+        if prefix.length == PREFIX_LEN:
+            return prefix_buf
+
+        if prefix.length > MAX_FRAME_SIZE:
+            raise CommError(CommError.BAD_DATA, f"Frame exceeds limit ({prefix.length} > {MAX_FRAME_SIZE}")
+
+        remaining = await self.reader.readexactly(prefix.length - PREFIX_LEN)
+
+        return prefix_buf + remaining
+
+    def _get_aio_properties(self) -> dict:
+
+        conn_props = {}
+        if not self.writer:
+            return conn_props
+
+        fileno = 0
+        local_addr = self.writer.get_extra_info("sockname", "")
+        if isinstance(local_addr, tuple):
```

Comment:
Under what circumstances will this `local_addr` be of other types?

Reply:
For TCP, it's a tuple with host and port. For UDS, it's just the file name.